README.md
# easyGL-ece297-2016
Improvements to EasyGL for use in UofT's ECE297 course.


changelog.txt
- added draw buffer
  - set draw target using set_drawing_buffer(ON_SCREEN || OFF_SCREEN)
  - push to screen using copy_off_screen_buffer_to_screen()
- moved FontCache to its own file fontcache
- moved t_x11_state and t_gl_state to their own file graphics_state
- added Cairo transparency in t_color
- added Surface class to load/draw pngs
  - Surface is its own class so users don't have to deal with
  initializing/freeing cairo surfaces
- added currentcoordinatesystem to gl_state to decide what coordinates the
draw functions draw to
  - GL_SCREEN turns off world_to_scrn translations


Surface.cpp
#include "Surface.h"

std::unordered_map<cairo_surface_t*, unsigned> Surface::surfaces;

Surface::Surface() {
    mSurface = NULL;
}

Surface::Surface(const char* filePath) {
    setSurface(filePath);
}

Surface::~Surface() {
    if (mSurface == NULL) return; // nothing to be done
    if (surfaces[mSurface] == 1)
        cairo_surface_destroy(mSurface);
    else if (surfaces[mSurface] < 1)
        std::cerr << "Yeah notify Vaughn that Harry screwed up again" << std::endl;
    --surfaces[mSurface];
}

// assignment operator
Surface& Surface::operator=(const Surface& surface) {
    mSurface = surface.mSurface;
    ++surfaces[mSurface];
    return *this;
}

// CCtor
Surface::Surface(const Surface& surface) {
    mSurface = surface.mSurface;
    ++surfaces[mSurface];
}

void Surface::setSurface(const char* filePath) {
    --surfaces[mSurface];
    cairo_surface_destroy(mSurface);
    mSurface = cairo_image_surface_create_from_png (filePath);
    ++surfaces[mSurface];
    switch(cairo_surface_status(mSurface)) {
        case CAIRO_STATUS_SUCCESS:
            return;
        case CAIRO_STATUS_NULL_POINTER:
            std::cerr << "Oops we messed up on our end." << std::endl << 
                    "Please notify Vaughn that Harry screwed up." << std::endl;
            break;
        case CAIRO_STATUS_NO_MEMORY:
            std::cerr << "Out of memory" << std::endl;
            break;
        case CAIRO_STATUS_READ_ERROR:
            std::cerr << "Error reading file" << std::endl;
            break;
        case CAIRO_STATUS_INVALID_CONTENT:
            std::cerr << "Invalid Content" << std::endl;
            break;
        case CAIRO_STATUS_INVALID_FORMAT:
            std::cerr << "Invalid Format" << std::endl;
            break;
        case CAIRO_STATUS_INVALID_VISUAL:
            std::cerr << "Invalid Visual" << std::endl;
            break;
        default:
            std::cerr << "Invalid Error" << std::endl;
            break;
    }

    // should only reach this point if unsuccessfull
    --surfaces[mSurface];
    cairo_surface_destroy(mSurface);
    mSurface = NULL;
}

cairo_surface_t* Surface::getSurface() const {
    return mSurface;
}


fontcache.h
#ifndef FONTCACHE_H
#define FONTCACHE_H

/**********************************
 * Common Preprocessor Directives *
 **********************************/
#include <stdlib.h>
#include <stdio.h>
#include <string>
#include <iostream>
#include <algorithm>
#include <vector>
#include <queue>
#include <unordered_map>
#include <sys/timeb.h>
#include <math.h>

#include "graphics.h"

#ifdef X11
#include <X11/Xlib.h>
#include <X11/Xutil.h>
#include <X11/Xos.h>
#include <X11/Xatom.h>
#include <X11/Xft/Xft.h>
#endif

#ifdef WIN32
#include <windows.h>
#include <WindowsX.h>
#endif

/**
 * Maximum sizes for the font cache. The "ZEROS" one is for fonts of zero rotation, and
 * the "ROTATED" one is for fonts with other rotation. If you plan on rotating text to
 * many rotations, say random ones, and displaying them all at the same time, I suggest
 * something large for "ROTATED". The value of "ZEROS" is probably fine.
 *
 * What these values really mean to you is: because the text drawing is really lazy about
 * loading fonts, the "ZEROS" should be at least how many font sizes you expect to be visible
 * at a given time, and the "ROTATED" is for how many other different angle-fontsize combinations
 * you expect to be visible at a given time. If you program remains within these limits, there
 * will be no slow frames, except the first one where you change many visible combinations.
 * For after this frame, the font cache has filled up with the fonts that are visible
 *
 * A quick experiment gave these results:
 *    10 fonts =>     540 KiB
 *   100 fonts =>   1.7   MiB
 *  1000 fonts =>  13.5   Mib
 * 10000 fonts => 132.4   Mib
 */
#define FONT_CACHE_SIZE_FOR_ZEROS 40
#define FONT_CACHE_SIZE_FOR_ROTATED 1000
#define NUM_FONT_TYPES 3

using namespace std;

#ifdef WIN32
typedef LOGFONT* font_ptr;
#endif

#ifdef X11
typedef XftFont* font_ptr;
#endif

/**
 * This is a class that caches fonts. It separates out unrotated and rotated fonts,
 * for the reason that the unrotated ones are used for bounds estimation anyway, as
 * well as that they are generally convenient to have around. For each group, it uses
 * a queue to keep track of the order of creation, and when a new font would put a
 * cache over capacity, it deletes the oldest one first. The maps are to allow fast
 * lookup of fonts from their size and rotation.
 *
 * This cache is also not very complicated, and it could be, if there were a good reason.
 *
 * Also note that no references retrieved from this cache should be retained, as
 * another part of the program may cause the creation of another font(s) that will
 * push yours out. Instead a font should be retrieved every time it is needed.
 * 
 * We're currently using only integer font rotations (in degrees), to avoid having
 * a tremendous number of possible font rotations.
 */
class FontCache {
public:
	FontCache()
	: order_zeros()
	, descriptor2font_zeros()
	, order_rotated()
	, descriptor2font_rotated() {
	}

	/**
	* Retrieve a font from this cache, or create it if it doesn't already exist,
	* pushing out, and deleting a font if the new font wont fit in its cache.
	* Gets a font from one of two caches depending on whether or not it is rotated.
	*/
	font_ptr get_font_info(size_t pointsize, int degrees);

	/**
	* Clear out this cache so that it contains no fonts. This deletes all fonts,
	* as well, so make sure any fonts that were retrieved before cannot be referenced.
	*/
	void clear();

private:
	typedef std::pair<size_t, int> font_descriptor;

	struct fontdesc_hasher {

	inline std::size_t operator()(const font_descriptor& v) const {
	    std::hash<size_t> sizet_hasher;
	    std::hash<int> int_hasher;
	    return sizet_hasher(v.first) ^ int_hasher(v.second);
	}
	};
	FontCache(const FontCache&);
	FontCache& operator=(const FontCache&);

	static void close_font(font_ptr font);
	static font_ptr do_font_loading(int pointsize, int degrees);

	// this function actually does all the work of get_font_info, but only
	// looks/creates in the given map and queue.
	template<class queue_type, class map_type>
	static font_ptr get_font_info(
	size_t pointsize, int degrees,
	queue_type& orderqueue, map_type& descr2font_map, size_t max_size);

	// unrotated fonts (zero rotation)
	std::deque<font_descriptor> order_zeros;
	std::unordered_map<font_descriptor, font_ptr, fontdesc_hasher> descriptor2font_zeros;
	// rotated fonts
	std::deque<font_descriptor> order_rotated;
	std::unordered_map<font_descriptor, font_ptr, fontdesc_hasher> descriptor2font_rotated;
};

#endif // FONTCACHE_H


fontcache.cpp
#include "fontcache.h"
#include "graphics_state.h"

/**
 * Linux/Mac:
 *   The strings listed below will be submitted to fontconfig, in this priority,
 *   with point size specified, and it will try to find something that matches. If it doesn't
 *   find anything, it will try the next font.
 *
 *   If fontconfig can't find any that look like any of the fonts you specify, (unlikely with the defaults)
 *   you must find one that it will. Run `fc-match <test_font_name>` to test you font name, and then modify
 *   this array with that name string. If you'd like, run `fc-list` to list available fonts.
 *   Note:
 *     - The fc-* commands are a CLI frontend for fontconfig,
 *       so anything that works there will work in this graphics library.
 *     - The font returned by fontconfig may not be the same, but will look similar.
 *
 * Windows:
 *   Something with similar effect will be done, but it is harder to tell what will happen.
 *   Check your installed fonts if the program exits immediately or while changing font size.
 *
 */
const char* const fontname_config[]{
    "helvetica",
    "lucida sans",
    "schumacher"
};

/******************************************
 * begin FontCache function definitions *
 ******************************************/

void FontCache::close_font(font_ptr font) {
#ifdef X11
    XftFontClose(t_x11_state::getInstance()->display, font);
#elif WIN32
    free(font);
#endif
}

/**
 * Loads the font with given attributes, and puts the pointer to in in put_font_ptr_here
 */
font_ptr FontCache::do_font_loading(int pointsize, int degrees) {

    bool success = false;
    font_ptr retval = NULL;

#ifdef X11
    for (int ifont = 0; ifont < NUM_FONT_TYPES; ifont++) {

#ifdef VERBOSE
        printf("Loading font: %s-%d\n", fontname_config[ifont], pointsize);
#endif

        XftMatrix font_matrix;
        XftMatrixInit(&font_matrix);
        if (degrees != 0) {
            XftMatrixRotate(&font_matrix, cos(PI * (degrees) / 180), sin(PI * degrees / 180));
        }

        /* Load font and get font information structure. */
        retval = XftFontOpen(
            t_x11_state::getInstance()->display, t_x11_state::getInstance()->screen_num,
            XFT_FAMILY, XftTypeString, fontname_config[ifont],
            XFT_SIZE, XftTypeDouble, (double) pointsize,
            XFT_MATRIX, XftTypeMatrix, &font_matrix,
            NULL // (sentinel)
            );

        if (retval == NULL) {
#ifdef VERBOSE
            fprintf(stderr, "Cannot open font %s", fontname_config[ifont]);
            if (degrees != 0) {
                fprintf(stderr, "with rotation %f deg", degrees);
            }
            printf("\n");
#endif
        } else {
            success = true;
            break;
        }
    }
    if (success == false) {
        printf("Error in load_font: fontconfig couldn't find any font of pointsize %d.\n", pointsize);
        if (degrees != 0) {
            printf("and rotation %d deg", degrees);
        }
        printf("Use `fc-list` to list available fonts, `fc-match` to test, and then modify\n");
        printf("the font config array in easygl_constants.h .\n");
        exit(1);
    }
#elif WIN32
    LOGFONT *lf = retval = (LOGFONT*) my_malloc(sizeof (LOGFONT));
    ZeroMemory(lf, sizeof (LOGFONT));
    // lfHeight specifies the desired height of characters in logical units.
    // A positive value of lfHeight will request a font that is appropriate
    // for a line spacing of lfHeight. On the other hand, setting lfHeight
    // to a negative value will obtain a font height that is compatible with
    // the desired pointsize.
    lf->lfHeight = -pointsize;
    lf->lfWeight = FW_NORMAL;
    lf->lfCharSet = ANSI_CHARSET;
    lf->lfOutPrecision = OUT_DEFAULT_PRECIS;
    lf->lfClipPrecision = CLIP_DEFAULT_PRECIS;
    lf->lfQuality = PROOF_QUALITY;
    lf->lfPitchAndFamily = VARIABLE_PITCH | FF_SWISS;
    lf->lfEscapement = (LONG) degrees * 10;
    lf->lfOrientation = (LONG) degrees * 10;
    HFONT testfont;
    for (int ifont = 0; ifont < NUM_FONT_TYPES; ++ifont) {
        MultiByteToWideChar(CP_UTF8, 0, fontname_config[ifont], -1,
            lf->lfFaceName, sizeof (lf->lfFaceName) / sizeof (wchar_t));

        testfont = CreateFontIndirect(lf);
        if (testfont == NULL) {
#ifdef VERBOSE
            fprintf(stderr, "Couldn't open font %s in pointsize %d.\n",
                fontname_config[ifont], pointsize);
#endif
            continue;
        }

        if (DeleteObject(testfont) == 0) {
            WIN32_DELETE_ERROR();
        } else {
            success = true;
            break;
        }
    }
    if (success == false) {
        printf("Error in load_font: Windows couldn't find any font of pointsize %d.\n", pointsize);
        printf("check installed fonts, and then modify\n");
        printf("the font config array in easygl_constants.h .\n");
        exit(1);
    }
#endif
    return retval;
}

void FontCache::clear() {
    for (
        auto iter = descriptor2font_zeros.begin();
        iter != descriptor2font_zeros.end();
        ++iter
        ) {
        close_font(iter->second);
    }
    order_zeros.clear();
    descriptor2font_zeros.clear();

    for (
        auto iter = descriptor2font_rotated.begin();
        iter != descriptor2font_rotated.end();
        ++iter
        ) {
        close_font(iter->second);
    }
    order_rotated.clear();
    descriptor2font_rotated.clear();
}

font_ptr FontCache::get_font_info(size_t pointsize, int degrees) {
    if (degrees == 0) {
        return get_font_info(
            pointsize, degrees,
            order_zeros, descriptor2font_zeros, FONT_CACHE_SIZE_FOR_ZEROS
            );
    } else {
        return get_font_info(
            pointsize, degrees,
            order_rotated, descriptor2font_rotated, FONT_CACHE_SIZE_FOR_ROTATED
            );
    }
}

template<class queue_type, class map_type>
font_ptr FontCache::get_font_info(
    size_t pointsize, int degrees,
    queue_type& orderqueue, map_type& descr2font_map, size_t max_size) {

    auto search_result = descr2font_map.find(std::make_pair(pointsize, degrees));
    if (search_result == descr2font_map.end()) {

        if (orderqueue.size() + 1 > max_size) {
            // if too many fonts, remove the oldest font from the cache.
            font_descriptor fontdesc_to_remove = orderqueue.back();
            auto font_to_remove = descr2font_map.find(fontdesc_to_remove);

            close_font(font_to_remove->second);

            descr2font_map.erase(font_to_remove);
            orderqueue.pop_back();
            puts("font cache overflow");
        }

        font_ptr new_font = do_font_loading(pointsize, degrees);
        font_descriptor new_font_desc = std::make_pair(pointsize, degrees);
        orderqueue.push_front(new_font_desc);
        descr2font_map.insert(std::make_pair(new_font_desc, new_font));
        return new_font;
    } else {
        return search_result->second;
    }
}




example.cpp
#include <iostream>
#include <cfloat> // for FLT_MAX
#include <chrono>
#include <thread>
#include <cstdlib>
#include <vector>
#include <string>
#include "graphics.h"

// Callbacks for event-driven window handling.
void drawscreen(void);
void act_on_new_button_func(void (*drawscreen_ptr) (void));
void act_on_button_press(float x, float y, t_event_buttonPressed event);
void act_on_mouse_move(float x, float y);
void act_on_key_press(char c, int keysym);

// A handy delay function for the animation example
void delay(long milliseconds);

// State variables for the example showing entering lines and rubber banding
// and the new button example.
static bool line_entering_demo = false;
static bool have_rubber_line = false;
static t_point rubber_pt; // Last point to which we rubber-banded.
static std::vector<t_point> line_pts; // Stores the points entered by user clicks.
static int num_new_button_clicks = 0;


// You can use any coordinate system you want.
// The coordinate system below matches what you have probably seen in math 
// (ie. with origin in bottom left).
// Note, that text orientation is not affected. Zero degrees will always be the normal, readable, orientation.
const t_bound_box initial_coords = t_bound_box(0, 0, 1000, 1000);


// Use this coordinate system for a conventional 
// graphics coordinate system (ie. with inverted y; origin in top left)
// const t_bound_box initial_coords = t_bound_box(0,1000,1000,0); 


#ifndef DRAW_BUFFER
#define DRAW_BUFFER
#endif

/**/

int main() {

    std::cout << "About to start graphics.\n";

    /**************** initialize display **********************/

    // Set the name of the window (in UTF-8), with white background.
    init_graphics("Some Example Graphics", WHITE); // you could pass a t_color RGB triplet instead

#ifdef DRAW_BUFFER
    set_drawing_buffer(OFF_SCREEN);
#endif

    // Set-up coordinates. The coordinates passed in define what coordinate
    // limits you want to use; this coordinate system is mapped to fill 
    // the screen.
    set_visible_world(initial_coords);

    // Set the message at the bottom (also UTF-8)
    update_message("Interactive graphics example.");

    // Pass control to the window handling routine.  It will watch for user input
    // and redraw the screen / pan / zoom / etc. the graphics in response to user
    // input or windows being moved around the screen.  We have to pass in 
    // at least one callback -- the one to redraw the graphics.
    // Three other callbacks can be provided to handle mouse button presses,
    // mouse movement and keyboard button presses in the graphics area, 
    // respectively. Those 3 callbacks are optional, so we can pass NULL if
    // we don't need to take any action on those events, and we do that
    // below.
    // This function will return if and when
    // the user presses the proceed button.

    event_loop(NULL, NULL, NULL, drawscreen);
    t_bound_box old_coords = get_visible_world(); // save the current view for later;

    /******************* animation section *******************************/
    // User might have panned and zoomed, which changes what world 
    // coordinates are visible on the screen. Get back to a known view.
    set_visible_world(initial_coords);
    clearscreen();
    update_message("Non-interactive (animation) graphics example.");
    setcolor(RED);
    setlinewidth(1);
    setlinestyle(DASHED);

    t_point start_point = t_point(100, 0);
    for (int i = 0; i < 50; i++) {
        drawline(start_point, start_point + t_point(500, 10));
        flushinput();
        delay(50);
        start_point += t_point(5, 10);
    }

    /**** Draw an interactive still picture again.  I'm also creating one new button. ****/

    update_message("Interactive graphics #2 Click in graphics area to rubber band line.");
    create_button("Window", "0 Clicks", act_on_new_button_func); // name is UTF-8

    // Enable mouse movement (not just button presses) and key board input.
    // The appropriate callbacks will be called by event_loop.
    set_keypress_input(true);
    set_mouse_move_input(true);
    line_entering_demo = true;

    // draw the screen once before calling event loop, so the picture is correct 
    // before we get user input.
    set_visible_world(old_coords); // restore saved coords -- this takes us back to where the user panned/zoomed.

    drawscreen();

    // Call event_loop again so we get interactive graphics again.
    // This time pass in all the optional callbacks so we can take
    // action on mouse buttons presses, mouse movement, and keyboard
    // key presses.
    event_loop(act_on_button_press, act_on_mouse_move, act_on_key_press, drawscreen);

    close_graphics();
    std::cout << "Graphics closed down.\n";

    return (0);
}

void drawscreen(void) {

    /* The redrawing routine for still pictures.  The graphics package calls  
     * this routine to do redrawing after the user changes the window 
     * in any way.                                                    
     */

    set_draw_mode(DRAW_NORMAL); // Should set this if your program does any XOR drawing in callbacks.
    clearscreen(); /* Should precede drawing for all drawscreens */

    setfontsize(10);
    setlinestyle(SOLID);
    setlinewidth(1);
    setcolor(BLACK);

    {
        /**************
         * Draw some rectangles using the indexed colours
         **************/

        color_types color_indicies[] = {
            LIGHTGREY,
            DARKGREY,
            WHITE,
            BLACK,
            BLUE,
            GREEN,
            YELLOW,
            CYAN,
            RED,
            DARKGREEN,
            MAGENTA
        };

        const float rectangle_width = 50;
        const float rectangle_height = rectangle_width;
        const t_point start_point = t_point(150, 30);
        t_bound_box color_rectangle = t_bound_box(start_point, rectangle_width, rectangle_height);

        // text is UTF-8
        drawtext(110, color_rectangle.get_ycenter(), "colors", 2 * (start_point.x - 110), rectangle_height);

        for (size_t i = 0; i < sizeof (color_indicies) / sizeof (color_indicies[0]); ++i) {
            setcolor(color_indicies[i]);
            fillrect(color_rectangle);
            color_rectangle += t_point(rectangle_width, 0);
        }

        drawtext(400, color_rectangle.get_ycenter(), "fillrect", FLT_MAX, rectangle_height);

        /********
         * Draw some rectangles with RGB triplet colours
         ********/

        std::srand((int) time(0)); // hack to make the colors change once per second.

        for (size_t i = 0; i < 3; ++i) {
            color_rectangle += t_point(rectangle_width, 0);
            setcolor(std::rand(), std::rand(), std::rand()); // note that setcolor(..) also takes a t_color object.
            fillrect(color_rectangle);
        }

        setcolor(BLACK);
        drawrect(start_point, color_rectangle.top_right());

    }

    /********
     * Draw some example lines, shapes, and arcs
     ********/
    {
        float radius = 50;

        setcolor(BLACK);
        drawtext(250, 150, "drawline", 150.0, FLT_MAX);
        setlinestyle(SOLID);
        drawline(200, 120, 200, 200);
        setlinestyle(DASHED);
        drawline(300, 120, 300, 200);

        setcolor(MAGENTA);
        drawtext(450, 160, "drawellipticarc", 150, FLT_MAX);
        drawellipticarc(550, 160, 30, 60, 90, 270);
        t_point cen(800, 160);
        drawtext(720, 160, "fillellipticarc", 150, FLT_MAX);
        fillellipticarc(cen, 30, 60, 90, 270);

        setcolor(BLUE);
        drawtext(190, 300, "drawarc", radius * 2, 150);
        drawarc(190, 300, radius, 0, 270);
        drawarc(300, 300, radius, 0, -180);
        fillarc(410, 300, radius, 90, -90);
        fillarc(520, 300, radius, 0, 360);
        setcolor(BLACK);
        drawtext(520, 300, "fillarc", radius * 2, 150);
        setcolor(BLUE);
        fillarc(630, 300, radius, 90, 180);
        fillarc(740, 300, radius, 90, 270);
        fillarc(850, 300, radius, 90, 30);
    }

    /********
     * Draw some rotated text
     ********/

    {
        const float textsquare_width = 200;

        t_bound_box textsquare = t_bound_box(t_point(100, 400), textsquare_width, textsquare_width);

        setcolor(BLUE);
        drawrect(textsquare);

        setcolor(GREEN);
        drawrect(textsquare.get_center(), textsquare.top_right());
        drawrect(textsquare.bottom_left(), textsquare.get_center());

        setcolor(RED);
        drawline(textsquare.bottom_left(), textsquare.top_right());
        drawline(textsquare.left(), textsquare.top(), textsquare.right(), textsquare.bottom());

        setcolor(t_color(0, 0, 0, 100));
        settextattrs(14, 0); // set a reasonable font size, and zero rotation.
        drawtext(textsquare.get_xcenter(), textsquare.bottom(), "0 degrees", textsquare.get_width(), textsquare.get_height());
        settextrotation(90);
        drawtext(textsquare.right(), textsquare.get_ycenter(), "90 degrees", textsquare.get_width(), textsquare.get_height());
        settextrotation(180);
        drawtext(textsquare.get_xcenter(), textsquare.top(), "180 degrees", textsquare.get_width(), textsquare.get_height());
        settextrotation(270);
        drawtext(textsquare.left(), textsquare.get_ycenter(), "270 degrees", textsquare.get_width(), textsquare.get_height());

        settextrotation(45);
        drawtext(textsquare.get_center(), "45 degrees", textsquare.get_width(), textsquare.get_height());
        settextrotation(135);
        drawtext(textsquare.get_center(), "135 degrees", textsquare.get_width(), textsquare.get_height());

        // It is probably a good idea to set text rotation back to zero,
        // as this is the way most text will be drawn, and you dont't want
        // to have to set it to zero every time you draw text. However,
        // there are checks in the easygl graphics that make it a no-op to set the text size, or
        // rotation, to the current value so there isn't a lot of speed hit to setting it all the time.
        settextrotation(0);
    }

    /*******
     * Draw some polygons
     *******/

    {
        t_point polypts[3] = {
            {500, 400},
            {450, 480},
            {550, 480}};
        t_point polypts2[4] = {
            {700, 400},
            {650, 480},
            {750, 480},
            {800, 400}};

        setfontsize(10);
        setcolor(RED);
        fillpoly(polypts, 3);
        fillpoly(polypts2, 4);
        setcolor(BLACK);
        drawtext(500, 450, "fillpoly1", 66.7, FLT_MAX);
        drawtext(725, 440, "fillpoly2", 100.0, FLT_MAX);
        setcolor(DARKGREEN);
        setlinestyle(SOLID);
        t_bound_box rect = t_bound_box(350, 550, 650, 670);
        drawtext_in(rect, "drawrect");
        drawrect(rect);
    }
    
    /*********
     * Draw some cairo things
     *********/
    {
        setfontsize(10);

        // Transparent lines
        for (int i = 1; i <= 2; ++i)
        {
            int offsetY = 50*i;
            if (i == 1) {setlinestyle(SOLID, ROUND); setcolor(t_color(0, 255, 0, 255/2));}
            else {setlinestyle(DASHED, BUTT); setcolor(t_color(255, 0, 0, 255/2));}
            drawtext(200, 900+offsetY, "Thin line (width 1)", 200.0, FLT_MAX);
            setlinewidth(1);
            drawline(100, 920+offsetY, 300, 920+offsetY);
            drawtext(500, 900+offsetY, "Width 3 Line", 200.0, FLT_MAX);
            setlinewidth(3);
            drawline(400, 920+offsetY, 600, 920+offsetY);
            drawtext(800, 900+offsetY, "Width 6 Line", 200.0, FLT_MAX);
            setlinewidth(6);
            drawline(700, 920+offsetY, 900, 920+offsetY);
        }
        
        // Transparent rectangles using fillrect and drawrect
        setlinestyle(SOLID);
        setlinewidth(3);
        setcolor(t_color(255, 0, 0, 255/2));
        fillrect(250, -100, 800, -50);
        drawrect(250, -100, 800, -50);
        setcolor(t_color(0, 0, 255, 255/2));
        fillrect(250, -100-50, 800, -50-50);
        drawrect(250, -100-50, 800, -50-50);
        setlinestyle(DASHED);
        setcolor(t_color(100, 200, 0, 255/2));
        fillrect(300, -100-25, 750, -50-25);
        drawrect(300, -100-25, 750, -50-25);
        
        // Polygons
        t_point polypts[4] = {
            {685, 380},
            {615, 500},
            {765, 500},
            {835, 380}};
        t_point polypts2[3] = {
            {500, 380},
            {425, 500},
            {575, 500}};
        setcolor(t_color(255, 100, 255, 255/2));
        fillpoly(polypts, 4);
        setcolor(t_color(100, 100, 255, 255/2));
        fillpoly(polypts2, 3);
        
        setlinewidth(1);
    }

    /*********
     * Do some things in screen coordinates
     *********/
    {
        set_coordinate_system(GL_SCREEN);
        setcolor(t_color(255, 0, 0));
        setlinestyle(SOLID);
        drawline(100, 100, 300, 300);
        set_coordinate_system(GL_WORLD);
    }
    
    /*********
     * Draw some example text, with the bounding box functions
     *********/

    {
        const float text_example_width = 800;

        const int num_lines = 2;
        const int max_strings_per_line = 3;

        const int num_strings_per_line[num_lines] = {3, 2};

        // Text is UTF-8, so you can use special characters, as long as your font supports them.
        const char* const line_text[num_lines][max_strings_per_line] = {
            {
                "8 Point Text",
                "12 Point Text",
                "18 Point Text"
            },
            {
                "24 Point Text",
                "32 Point Text"
            }
        };

        const int text_sizes[num_lines][max_strings_per_line] = {
            {8, 12, 15},
            {24, 32}
        };

        setcolor(BLACK);
        setlinestyle(DASHED);

        for (int i = 0; i < num_lines; ++i) {
            t_bound_box text_bbox = t_bound_box(t_point(100., 710. + i * 60.), text_example_width / num_strings_per_line[i], 60.);
            for (int j = 0; j < num_strings_per_line[i]; ++j) {
                setfontsize(text_sizes[i][j]);
                drawtext_in(text_bbox, line_text[i][j]);
                drawrect(text_bbox);

                text_bbox += t_point(text_example_width / num_strings_per_line[i], 0);
            }
        }
    }

    /********
     * Draw some lines of various thickness
     ********/
    {
        setfontsize(10);

        setlinestyle(SOLID);
        drawtext(200, 900, "Thin line (width 1)", 200.0, FLT_MAX);
        setlinewidth(1);
        drawline(100, 920, 300, 920);
        drawtext(500, 900, "Width 3 Line", 200.0, FLT_MAX);
        setlinewidth(3);
        setlinestyle(SOLID, ROUND);
        drawline(400, 920, 600, 920);
        drawtext(800, 900, "Width 6 Line", 200.0, FLT_MAX);
        setlinewidth(6);
        drawline(700, 920, 900, 920);
    }

    /********
     * Draw the rubber-band line, if it's there
     ********/

    setlinewidth(1);
    setcolor(GREEN);

    for (unsigned int i = 1; i < line_pts.size(); i++)
        drawline(line_pts[i - 1], line_pts[i]);

    // Screen redraw will get rid of a rubber line.  
    have_rubber_line = false;

    /********
     * Draw a small png
     ********/
    draw_surface(load_png_from_file("small_image.png"), 100, 100);

#ifdef DRAW_BUFFER
    copy_off_screen_buffer_to_screen();
#endif
}

void delay(long milliseconds) {
    // if you would like to use this function in your
    // own code you will need to #include <chrono> and
    // <thread>
    std::chrono::milliseconds duration(milliseconds);
    std::this_thread::sleep_for(duration);
}

void act_on_new_button_func(void (*drawscreen_ptr) (void)) {

    char old_button_name[200], new_button_name[200];
    std::cout << "You pressed the new button!\n";
    setcolor(MAGENTA);
    setfontsize(12);
    drawtext(500, 500, "You pressed the new button!", 10000.0, FLT_MAX);
    sprintf(old_button_name, "%d Clicks", num_new_button_clicks);
    num_new_button_clicks++;
    sprintf(new_button_name, "%d Clicks", num_new_button_clicks);
    change_button_text(old_button_name, new_button_name);

    // Re-draw the screen (a few squares are changing colour with time)
    drawscreen_ptr();
}

void act_on_button_press(float x, float y, t_event_buttonPressed event) {

    /* Called whenever event_loop gets a button press in the graphics *
     * area.  Allows the user to do whatever he/she wants with button *
     * clicks.                                                        */

    std::cout << "User clicked a mouse button at coordinates ("
            << x << "," << y << ")";
    if (event.shift_pressed || event.ctrl_pressed) {
        std::cout << " with ";
        if (event.shift_pressed) {
            std::cout << "shift ";
            if (event.ctrl_pressed)
                std::cout << "and ";
        }
        if (event.ctrl_pressed)
            std::cout << "control ";
        std::cout << "pressed.";
    }
    std::cout << std::endl;

    if (line_entering_demo) {
        line_pts.push_back(t_point(x, y));
        have_rubber_line = false;

        // Redraw screen to show the new line.  Could do incrementally, but this is easier.
        drawscreen();
    }
}

void act_on_mouse_move(float x, float y) {
    // function to handle mouse move event, the current mouse position in the current world coordinate
    // system (as defined in your call to init_world) is returned

    std::cout << "Mouse move at " << x << "," << y << ")\n";
    if (line_pts.size() > 0) {
        // Rubber banding to a previously entered point.
        // Go into XOR mode.  Make sure we set the linestyle etc. for xor mode, since it is 
        // stored in different state than normal mode.
        set_draw_mode(DRAW_XOR);
        setlinestyle(SOLID);
        setcolor(WHITE);
        setlinewidth(1);
        int ipt = line_pts.size() - 1;

        if (have_rubber_line) {
            // Erase old line.  
            drawline(line_pts[ipt], rubber_pt);
        }
        have_rubber_line = true;
        rubber_pt.x = x;
        rubber_pt.y = y;
        drawline(line_pts[ipt], rubber_pt); // Draw new line
    }
}


#include <X11/keysym.h>

void act_on_key_press(char c, int keysym) {
    // function to handle keyboard press event, the ASCII character is returned
    // along with an extended code (keysym) on X11 to represent non-ASCII
    // characters like the arrow keys.

    std::cout << "Key press: char is " << c << std::endl;

#ifdef X11 // Extended keyboard codes only supported for X11 for now
    switch (keysym) {
        case XK_Left:
            std::cout << "Left Arrow" << std::endl;
            break;
        case XK_Right:
            std::cout << "Right Arrow" << std::endl;
            break;
        case XK_Up:
            std::cout << "Up Arrow" << std::endl;
            break;
        case XK_Down:
            std::cout << "Down Arrow" << std::endl;
            break;
        default:
            std::cout << "keysym (extended code) is " << keysym << std::endl;
            break;
    }
#endif
}

